Allow assigning ellipsis literal as parameter default value (#14982)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/stubs/ellipsis.md

Reviewer: Could you possibly also add a test snippet demonstrating that `...` is _not_ allowed as a default value in a non-stub file if it contradicts the annotation?

# Ellipsis

## Function and Methods

For default values the ellipsis literal `...` can be used.

```py
from typing import Dict

def f(x: int = ...) -> None: ...

> def f2(x: Dict = ...) -> None: ...

Author: Thanks for the suggestion. I added it.

# Ellipsis

## Function and Methods

> For default values the ellipsis literal `...` can be used.
> 
> ```py

Reviewer 2: This special case for `...` default values is only allowed in stub files, and the default mdtest path is `test.py`, which is not a stub file. So every test showing the special case working should have an explicit file path with `pyi` prefix, and we should also have a test for a `.py` filename showing that the special case is not applied, and a diagnostic for invalid default-value type is emitted.
```suggestion
For default values the ellipsis literal `...` can be used, in a stub file only.

```py path=test.pyi
```

> from typing import Dict
> 
> def f(x: int = ...) -> None: ...
> def f2(x: Dict = ...) -> None: ...

Reviewer 2: importing `typing.Dict` seems like an extraneous wrinkle for these tests. since there's no special behavior around it here.
```suggestion
def f(x: int = ...) -> None: ...
def f2(x: dict = ...) -> None: ...
```

> 
> # error: [invalid-assignment] "Object of type `EllipsisType | ellipsis` is not assignable to `int`"
> y: int = ...

Reviewer 2: This seems redundant with the dedicated "Class and Module Level Attributes" test below.
```suggestion
```

```

## Use of Ellipsis Symbol

When the ellipsis symbol is used as default value the assignment is checked.

> ```py

Reviewer 2: ```suggestion
```py path=test.pyi
```

> from typing import Dict
> 

Reviewer 2: ```suggestion
```

# error: [invalid-parameter-default] "Default value of type `EllipsisType | ellipsis` is not assignable to annotated parameter type `int`"
def f(x: int = Ellipsis) -> None: ...
```

## Class and Module Level Attributes

Using ellipsis literal for classes and module level attributes is unnecessary and results in an
error.

- <https://typing.readthedocs.io/en/latest/guides/writing_stubs.html#module-level-attributes>

> ```py

Reviewer 2: ```suggestion
```py path=test.pyi
```

# Ellipsis

## Function and methods

The ellipsis literal `...` can be used as a placeholder default value for a function parameter, in a
stub file only, regardless of the type of the parameter.

```py path=test.pyi
def f(x: int = ...) -> None:
    reveal_type(x)  # revealed: int

def f2(x: str = ...) -> None:
    reveal_type(x)  # revealed: str
```

> ## Class and module level attributes
> 
> The ellipsis literal can be assigned to a class or module attribute, regardless of its type, in a
> stub file only.

Reviewer: ```suggestion
## Class and module symbols

The ellipsis literal can be assigned to a class or module symbol, regardless of its declared type,
in a stub file only.
```

```py path=test.pyi
y: bytes = ...
reveal_type(y)  # revealed: bytes
x = ...
reveal_type(x)  # revealed: Unknown

class Foo:
    y: int = ...

reveal_type(Foo.y)  # revealed: int
```

## Unpacking ellipsis literal in assignment

> 

Reviewer: ```suggestion

No diagnostic is emitted if an ellipsis literal is "unpacked" in a stub file as part of an
assignment statement:

```

```py path=test.pyi
x, y = ...
reveal_type(x)  # revealed: Unknown
reveal_type(y)  # revealed: Unknown
```

## Unpacking ellipsis literal in for loops

> 

Reviewer: ```suggestion

Iterating over an ellipsis literal as part of a `for` loop in a stub is invalid, however, and
results in a diagnostic:

```

```py path=test.pyi
# error: [not-iterable] "Object of type `ellipsis` is not iterable"
for a, b in ...:
    reveal_type(a)  # revealed: Unknown
    reveal_type(b)  # revealed: Unknown
```

## Ellipsis usage in non stub file

In a non-stub file, there's no special treatment of ellipsis literals. An ellipsis literal can only
be assigned if `EllipsisType` is actually assignable to the annotated type.

```py
# error: 7 [invalid-parameter-default] "Default value of type `ellipsis` is not assignable to annotated parameter type `int`"
def f(x: int = ...) -> None: ...

# error: 1 [invalid-assignment] "Object of type `ellipsis` is not assignable to `int`"
a: int = ...
b = ...
reveal_type(b)  # revealed: ellipsis
```

> ## Use of ellipsis symbol

Reviewer: ```suggestion
## Use of `Ellipsis` symbol
```

> There is no special treatment of the builtin name `Ellipsis`, only of `...` literals.

Reviewer: ```suggestion
There is no special treatment of the builtin name `Ellipsis` in stubs, only of `...` literals.
```
